petastorm/pytorch.py

```diff
@@ -213,3 +214,106 @@ def __enter__(self):
+        returned to a user by the ``BatchedDataLoader`` until the queue is full. After that, batches of `batch_size`
+        will be created by uniformly sampling the shuffling queue. Once no more samples are available from the data
+        reader, the shuffling queue is allowed to be consumed till no further samples are available.
+
+        Note that the last returned batch could have less then ``batch_size`` samples.
+
+        NOTE: if you are using ``make_batch_reader``, this shuffling queue will be randomizing the order of the
+        entire batches and not changing the order of elements within a batch. This is likely not what you intend to do.
+
+        Unlike `DataLoader`, this class does not support special types that are not supported in PyTorch (decimal/string).
+
+        :param reader: petastorm Reader instance
+        :param batch_size: the number of items to return per batch; factored into the len() of this reader
+        :param transform_fn: an optional callable to convert batches from the reader to PyTorch tensors
+        :param shuffling_queue_capacity: Queue capacity is passed to the underlying :class:`tf.RandomShuffleQueue`
+          instance. If set to 0, no suffling will be done.
+        """
+        self.reader = reader
+        self.batch_size = batch_size
+        self.transform_fn = transform_fn
+
+        # _batch_acc accumulates samples for a single batch.
+        self._batch_acc = []
+        if shuffling_queue_capacity > 0:
+            # We can not know what is the reasonable number to use for the extra capacity, so we set a huge number
+            # and give up on the unbound growth protection mechanism.
+            min_after_dequeue = shuffling_queue_capacity - 1
+            self._shuffling_buffer = BatchedRandomShufflingBuffer(
+                shuffling_queue_capacity,
+                min_after_retrieve=min_after_dequeue,
+                extra_capacity=100000000,
+                batch_size=batch_size
+            )
+        else:
+            self._shuffling_buffer = BatchedNoopShufflingBuffer(batch_size=batch_size)
+
+    def __iter__(self):
+        """
+        The Data Loader iterator stops the for-loop when reader runs out of samples.
+        """
+        # As we iterate over incoming samples, we are going to store them in `self._batch_acc`, until we have a batch of
+        # the requested batch_size ready.
+
+        keys = None
+
+        for row in self.reader:
+            # Default collate does not work nicely on namedtuples and treat them as lists
+            # Using dict will result in the yielded structures being dicts as well
+            row_as_dict = row._asdict()
+
+            keys = row_as_dict.keys()
+
+            # Promote some types that are incompatible with pytorch to be pytorch friendly.
+            _sanitize_pytorch_types(row_as_dict)
+
+            # Add rows to shuffling buffer
```

Comment:
Wouldn't it be faster to perform this load into shuffling buffer in a separate thread rather than doing it in the same read loop that returns the batch?

Reply:
The add_many and retrieve functions are not thread-safe. In a future version, we can add queues to get better performance but I figured this was the MVP to get things working.

Reply:
makes sense.